File: LeetCode/1143. Longest Common Subsequence/solution.cpp
/**
 * 44 / 44 test cases passed.
 * Runtime: 24 ms
 * Memory Usage: 12.8 MB 
 */
class Solution {
public:
    int longestCommonSubsequence(string text1, string text2) {
        vector<vector<int>> dp(text1.size() + 1, vector<int>(text2.size() + 1));
        for (int i = 1; i <= text1.size(); ++ i) {
            for (int j = 1; j <= text2.size(); ++ j) {
                dp[i][j] = text1[i - 1] == text2[j - 1] ? dp[i - 1][j - 1] + 1 : max(dp[i - 1][j], dp[i][j - 1]);
            }
        }
        return dp.back().back();
    }
};

File: LeetCode/1143. Longest Common Subsequence/solution.py
"""
44 / 44 test cases passed.
Runtime: 468 ms
Memory Usage: 22.4 MB
"""
class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        sz1, sz2 = len(text1), len(text2)
        dp = [[0] * (sz2 + 1) for _ in range(sz1 + 1)]
        for i in range(1, sz1 + 1):
            for j in range(1, sz2 + 1):
                dp[i][j] = dp[i-1][j-1] + 1 if text1[i-1] == text2[j-1] else max(dp[i-1][j], dp[i][j-1])
        return dp[-1][-1]
